### PR TITLE
sys/color: extend unittest and fix module

### PR DESCRIPTION
**Pull request by kfessel**
### Contribution description

this extends the unittest for sys_color testing more colors 

### Testing procedure

```
RIOT_tree/tests/unittests$ make tests-color test
```
will fail since our `rgb2hsv` implementation is wrong (or is using an other colorspace than hsv2rgb (without documenting))

the new `hsv2rgb` test will succeed

### Issues/PRs references

#19614  was the reason i had a look at this 

#1315 added the rgb2hsv and hsv2rgb function

#9940 added the test for black special case

https://www.vagrearg.org/content/hsvrgb << some optimization for that function (avoiding float)

### CODE REVIEW

**Comment by miri64** [comment]
(not sure what you mean by "wrong" so I provided the black-corner case thingy for documentation)

**Comment by kfessel** [comment]
The test previously tested only one value (black) (for good reason).
I extended the test to test more values (using one of the very similar common tables of typical values that has some hsv an rgb value examples) .
Sadly #9939 does not fix fix enough of that function (maybe there is a color space adjustment that is not very common in that function (but the hsv2rgb the common color space so i assume the difference in rgb2hsv is  an error).

File: sys/color/color.c [hunk 1]
```diff
@@ -27,51 +27,49 @@
...
 
     /* norm RGB colors to the range [0 - 1.0] */
-    rd = (float)rgb->r / 255.0f;
-    gd = (float)rgb->g / 255.0f;
-    bd = (float)rgb->b / 255.0f;
-
-    /* find value as maximum of the three colors */
-    if (rd >= gd) {
-        hsv->v = (rd >= bd) ? rd : bd;
+    rd = (float)rgb->r * (1 / 255.0f);
```
Comment (maribu) [inline sys/color/color.c:34]:
```suggestion
    /* multiplication with the compile-time constant `1 / 255.0f` is
     * faster than division by `255.0f` */
    rd = (float)rgb->r * (1 / 255.0f);
```

File: sys/color/color.c [hunk 2]
```diff
@@ -27,51 +27,49 @@
...
+    gd = (float)rgb->g * (1 / 255.0f);
+    bd = (float)rgb->b * (1 / 255.0f);
+
+    /* find maximum and minimum and quadrant of the three colors*/
+    imax = rgb->r;
+    max = rd;
+    q = 0;
+    if ( rgb->g > imax ) {
```
Comment (maribu) [inline sys/color/color.c:42]:
```suggestion
    if (rgb->g > imax) {
```

File: sys/color/color.c [hunk 3]
```diff
@@ -27,51 +27,49 @@
...
+        imax = rgb->g;
+        max = gd;
+        q = 1;
     }
-    else {
-        hsv->v = (gd >= bd) ? gd : bd;
+    if ( rgb->b > imax ) {
```
Comment (maribu) [inline sys/color/color.c:47]:
```suggestion
    if (rgb->b > imax) {
```

File: sys/color/color.c [hunk 4]
```diff
@@ -27,51 +27,49 @@
...
 
     /* compute delta from value and minimum of the three RGB colors */
-    if (rd <= gd) {
-        delta = (rd <= bd) ? (hsv->v - rd) : (hsv->v - bd);
-    }
-    else {
-        delta = (gd <= bd) ? (hsv->v - gd) : (hsv->v - bd);
-    }
+    imin = rgb->r, min = rd;
+    imin = (rgb->g < imin) ? (min = gd , rgb->g) : imin;
```
Comment (maribu) [inline sys/color/color.c:57]:
```suggestion
    imin = rgb->r;
    min = rd;
   if (rgb->g < imin) {
       imin = rgb->g;
       min = gd;
   }
```

File: sys/color/color.c [hunk 5]
```diff
@@ -27,51 +27,49 @@
...
+    imin = (rgb->b < imin) ? (min = bd , rgb->b) : imin;
```
Comment (maribu) [inline sys/color/color.c:58]:
```suggestion
    if (rgb->b < imin) {
        imin =  rgb->b;
        min = bd;
    }
```

File: sys/color/color.c [hunk 6]
```diff
@@ -27,51 +27,49 @@
...
-            hsv->h = 2.0f + rc - bc;
-        }
-        else {
-            hsv->h = 4.0f + gc - rc;
+    if (min != max) {
+        float p = 60.0f / delta;
+        switch (q){
+            case 0: hsv->h = (gd - bd) * p; break;
+            case 1: hsv->h = 120.0f + (bd - rd) * p; break;
+            case 2: hsv->h = 240.0f + (rd - gd) * p; break;
```
Comment (maribu) [inline sys/color/color.c:71]:
```suggestion
        switch (q) {
        case 0:
            hsv->h = (gd - bd) * p;
            break;
        case 1:
            hsv->h = 120.0f + (bd - rd) * p;
            break;
        case 2:
            hsv->h = 240.0f + (rd - gd) * p;
            break;
```

File: tests/unittests/tests-color/tests-color.c [hunk 7]
```diff
@@ -75,20 +75,92 @@ static void test_rgb2hex__success(void)
     TEST_ASSERT_EQUAL_INT(0x000AB13C, hex);
 }
 
-static void test_rgb2hsv__black(void)
+
```
Comment (maribu) [inline tests/unittests/tests-color/tests-color.c:78]:
```suggestion
```

File: sys/color/color.c (hunk 4 above)
Comment (kfessel) [reply sys/color/color.c:57]: don't you like the comma operator?

Comment (maribu) [reply sys/color/color.c]: No. Many people have to look up its meaning again to understand the code ;-)

File: tests/unittests/tests-color/tests-color.c [hunk 8]
```diff
@@ -75,20 +75,91 @@ static void test_rgb2hex__success(void)
...
+        { { 180, 100, 100 }, { 0, 255, 255 } }, /*Cyan*/
+        { { 300, 100, 100 }, { 255, 0, 255 } }, /*Magenta*/
+        { { 0, 0, 75 }, { 191, 191, 191 } },    /*Silver*/
+        { { 0, 0, 50 }, { 128, 128, 128 } },    /*Gray*/
+        { { 0, 100, 50 }, { 128, 0, 0 } },      /*Maroon*/
+        { { 60, 100, 50 }, { 128, 128, 0 } },   /*Olive*/
+        { { 120, 100, 50 }, { 0, 128, 0 } },    /*Green*/
+        { { 300, 100, 50 }, { 128, 0, 128 } },  /*Purple*/
+        { { 180, 100, 50 }, { 0, 128, 128 } },  /*Teal*/
+        { { 240, 100, 50 }, { 0, 0, 128 } }/*Navy*/
```
Comment (maribu) [inline tests/unittests/tests-color/tests-color.c:143]:
```suggestion
        { { 120, 100, 50 }, { 0, 128, 0 } },    /*Green*/
        { { 300, 100, 50 }, { 128, 0, 128 } },  /*Purple*/
        { { 180, 100, 50 }, { 0, 128, 128 } },  /*Teal*/
        { { 240, 100, 50 }, { 0, 0, 128 } }     /*Navy*/
```

File: tests/unittests/tests-color/tests-color.c [hunk 9]
```diff
@@ -75,20 +75,91 @@ static void test_rgb2hex__success(void)
...
+    };
+    unsigned len = sizeof h_r / sizeof h_r[0];
```
Comment (maribu) [inline tests/unittests/tests-color/tests-color.c:145]:
```suggestion
    unsigned len = ARRAY_SIZE(h_r);
```

**Review by maribu — APPROVED** [review]
The code looks good to me and decent test coverage was added.

Please squash the minor style nitpicks directly; the ACK remains valid.

**Comment by kfessel** [comment]
bors merge

**Comment by kfessel** [comment]
bors retry

**Comment by kfessel** [comment]
bors cancel 
bors merge

**Comment by maribu** [comment]
bors merge

**Comment by maribu** [comment]
bors merge

**Pull request merged**